src/main/java/com/cloudbees/jenkins/support/filter/WordsTrie.java

```diff
@@ -0,0 +1,171 @@
+package com.cloudbees.jenkins.support.filter;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+
+/**
+ * Trie implementation to generate a "Trie regex". A regex that reduces backtracking by following a Trie structure.
+ * <p>
+ * When searching for a match within a list of words, for example ["go", "goes", "going", "gone", "goose"],  a simple
+ * regex that matches any word would typically look like {code}\b(?:(go|goes|going|gone|goose))\b{code}.
+ * <p>
+ * While this works, Such a pattern can be optimized significantly by following a Trie structure in the prefixes such
+ * as {code}\b(?:go(?:(?:es|ing|ne|ose))?)\b{code}.
+ */
+public class WordsTrie {
+
+    /*
+     * As per https://docs.oracle.com/javase/tutorial/essential/regex/literals.html, metacharacters need to be escaped
+     * in the regex String. The regular expression matches any metacharacter occurrence in a String, in which case
+     * the String will need to be quoted.
+     *
+     * NOTE: We could instead choose to escape everything and not do that check. This would increase the size of the
+     * regex String. This does seem to have a significant impact on performance. So escaping only those characters for
+     * now.
+     */
+    private final static Pattern METACHARACTER =
+        Pattern.compile("[\\x21\\x24\\x28-\\x2B\\x2D-\\x2F\\x3C-\\x3F\\x5B-\\x5E\\x7B-\\x7D]+");
+
+    final TrieNode root;
+
+    public WordsTrie() {
+        this.root = new TrieNode(false);
+    }
+
+    /**
+     * Add a word to the Trie.
+     * @param word the word
+     */
+    public void add(String word) {
+        TrieNode ref = root;
+        int i = 0;
+        while (i < word.length() - 1) {
+            // Fill in down the Trie
+            ref = ref.getOrCreate(word.charAt(i), s -> new TrieNode(false));
+            i++;
+        }
+        // We need to mark the last node as an end node
+        ref.getOrCreate(word.charAt(i), s -> new TrieNode(true)).end = true;
```

Comment:
I doubt it matters in this context, but beware emojis which would have multiple `char`s per code point.

Reply:
Maybe I can write a test for all chars. Like in https://github.com/jenkinsci/support-core-plugin/blob/1283.v9ddb_0284a_00c/src/test/java/com/cloudbees/jenkins/support/util/WordReplacerTest.java#L76-L81.

Reply:
Interesting. Java Pattern with CASE_INSENSITIVE does not hanlde some of those chars. For example the following fails:

```
assertTrue(Pattern.compile("\\baİz\\b", Pattern.CASE_INSENSITIVE).matcher("ai̇z").matches());
```

`İ` has a single char but its lowercase value `i̇` has 2 chars..

Reply:
Eventually found a way to fix this with `UNICODE_CASE` and using the simpler `Matcher#replaceAll`.